File: src/java/magma/rule/ExactRule.java
package magma.rule;

import magma.Node;
import magma.error.CompileError;
import magma.error.StringContext;
import magma.result.Err;
import magma.result.Ok;
import magma.result.Result;

public record ExactRule(String slice) implements Rule {
    @Override
    public Result<Node, CompileError> parse(String input) {
        if (input.equals(slice)) return new Ok<>(new Node());
        final var format = "Input does not match exactly '%s'";
        final var message = format.formatted(slice);
        final var context = new StringContext(input);
        return new Err<>(new CompileError(message, context));
    }

    @Override
    public Result<String, CompileError> generate(Node node) {
        return new Ok<>(slice);
    }
}

File: src/c/magma/app/rule/ExactRule.c
import magma.api.result.Err;import magma.api.result.Ok;import magma.api.result.Result;import magma.app.MapNode;import magma.app.Node;import magma.app.error.CompileError;import magma.app.error.context.StringContext;struct ExactRule(String slice){
	Result<Node, CompileError> parse(any* _ref_, String input){
		if(input.equals(this.slice))return new Ok<>(new MapNode());
		var context=new StringContext(input);
		return new Err<>(new CompileError("Exact string '"+this.slice + "' was not present", context));
	}
	Result<String, CompileError> generate(any* _ref_, Node node){
		return new Ok<>(this.slice);
	}
	Rule N/A(){
		return N/A.new();
	}
}